File: 2017/14.py

import math, copy, re

RANGE = 256
ROUNDS = 64
APPENDIX = [17, 31, 73, 47, 23]
GRID = []
VISITED = '2'
nr = 128
def sparse_hash(len_seq, rounds, LIST):
	skip = pos = 0
	for i in range(rounds):
		for L in len_seq:
			if pos + L > RANGE:
				sublist = LIST[pos :] + LIST[: (pos + L) % RANGE]
			else:
				sublist = LIST[pos : pos + L]
			sublist.reverse()
			if pos + L > RANGE:
				LIST[pos :] = sublist[: RANGE - pos]
				LIST[: (pos + L) % RANGE] = sublist[RANGE - pos :]
			else:
				LIST[pos : pos + L] = sublist
			pos = (pos + L + skip) % RANGE
			skip += 1

	return LIST

def dense_hash(lst):
	ret_hash = ''
	xor = 0
	L = len(lst)
	for i in range(L + 1):
		if i > 0 and i % 16 == 0:
			tmp = hex(xor)[2:]
			if len(tmp) == 1:
				tmp = '0' + tmp
			ret_hash += tmp
			xor = 0
		if i == L:
			break
		xor ^= lst[i]

	return ret_hash

def calc_bits(hsh, cout = False):
	ret = 0
	L = 32 - len(hsh)
	hsh = hsh + L * '0' 
	new_row = ""
	for char in hsh:
		ret += bin(int('0x' + char,16))[2:].count('1')
		s = bin(int('0x' + char,16))[2:]
		s = (4 - len(s)) * '0' + s
		new_row += s
	# new_row += L * '0'
	if cout:
		# print(hsh, len(hsh))
		print(new_row)

	GRID.append(list(new_row))
	return ret

def part_1(data):
	bits = 0
	for i in range(128):
		data_str = data + '-' + str(i)
		LIST = [j for j in range(RANGE)]
		len_seq = [ord(x) for x in data_str] + APPENDIX
		LIST = sparse_hash(len_seq, ROUNDS, LIST)
		hsh = dense_hash(LIST)
		bits += calc_bits(hsh)
	
	print(bits)

	print('END OF PART1')
	return

def dfs(r, c):
    GRID[r][c] = VISITED
    nc = len(GRID[r])
    if r - 1 >= 0 and len(GRID[r-1]) > c and GRID[r-1][c] == '1':
	    dfs(r-1, c)
    if r + 1 < nr and len(GRID[r+1]) > c and GRID[r+1][c] == '1':
      dfs(r+1, c)
    if c - 1 >= 0 and GRID[r][c-1] == '1':
        dfs(r, c-1)
    if c + 1 < nc and GRID[r][c+1] == '1':
        dfs(r, c+1)

def part_2(data):
	bits = 0
	for i in range(128):
		data_str = data + '-' + str(i)
		LIST = [j for j in range(RANGE)]
		len_seq = [ord(x) for x in data_str] + APPENDIX
		LIST = sparse_hash(len_seq, ROUNDS, LIST)
		hsh = dense_hash(LIST)
		calc_bits(hsh)

	islands = 0
	for row in range(len(GRID)):
		print(''.join(GRID[row]))
		for col in range(len(GRID[row])):
			if GRID[row][col] == '1':
				islands += 1
				dfs(row, col)

	print(islands)	
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('14_input') as f:
		data = f.read()
		# data = data.split('\n')
		# data = list(map(int, data.split()))


	# part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2017/14.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

const knotHash = data => {
	const dense = (L) => {
		const res = [];
		let curr = 0;
		range(L.length)(i => {
			if (i && i % 16 === 0) {
				res.push(curr);
				curr = 0;
			}
			curr ^= L[i];
		})
		res.push(curr);
		return res;
	}
	const reverse = (L, curr, n) => {
		const nArr = [];
		range(n)(i => {
			nArr.push(L[(curr + i) % L.length]);
		})
		nArr.reverse();
		range(n)(i => {
			L[(curr + i) % L.length] = nArr[i];
		})
	}
	const paddedHex = n => {
		n = hex(n);
		return n.length === 2 ? n : ('0' + n);
	}

	let res = 0;
	data = data.split('').map(c => ord(c)).concat([17, 31, 73, 47, 23])
	const LIST = new Array(256).fill(0).map((_, i) => i);
	let curr = 0;
	let skip = 0;

	range(64)(_ => {
		iter(data, n => {
			reverse(LIST, curr, n);
			curr = (curr + n + skip) % LIST.length;
			++skip; 
		})
	})
	const nList = dense(LIST);

	res = nList.map(n => paddedHex(n)).join('');
	return res;
}

function part1(data) {

	let res = 0;
	range(128)(i => {
		const hash = knotHash(`${data}-${i}`);
		iter(hash, hx => {
			const n = bin(num(hx, 16));
			iter(n, bit => {
				res += bit === '1';
			})
		})
	})

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

const fill = (G, i, j) => {
	const q = new Queue();
	q.push([i, j]);
	while (!q.empty()) {
		const [y, x] = q.pop();
		if (G[y][x] !== '1') continue;
		G[y][x] = '2';
		iter(D4, ([dy, dx]) => {
			if (inBB(y+dy, x+dx, G)) {
				q.push([y+dy, x+dx]);
			}
		})
	}
}

const calcGroups = G => {
	let res = 0;
	drange(G.length)((i, j) => {
		if (G[i][j] === '1') {
			// debug(i, j, res)
			// if (res%100===0) disp(G)
			fill(G, i, j);
			++res;
		}
	})
	return res;
}

function part2(data) {

	let res = 0;
	const G = [];
	range(128)(i => {
		let hash = knotHash(`${data}-${i}`);
		let s = '';
		iter(hash, hx => {
			let n = bin(num(hx, 16));
			while (n.length < 4) {
				n = '0' + n;
			}
			s += n;
		})
		G.push(s.split(''));
	})

	res = calcGroups(G);

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("14_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
